### PR TITLE
Add support for Jackson's JsonPOJOBuilder

### PR DESCRIPTION
**Pull request by misl**
Added support to Jackson extension for immutable POJO models using @JsonPOJOBuilder.

- Both model and builder are added as reflective classes
- Also supports model and builder inheritance (add superclasses as well)
- Added exclusion for @JsonDeserialize without builder field
- Includes integration tests.

### CODE REVIEW

**Comment by geoand** [comment]
A NPE is being thrown as a result of the changes introduced by the PR

**Comment by geoand** [comment]
@misl when you are done working on this, please squash your commits into a single one - thanks!

**Comment by geoand** [comment]
Also, I think that the PR needs to be rebased on `master` and update the native image CI configuration now that https://github.com/quarkusio/quarkus/pull/3719 has been merged

**Comment by misl** [comment]
> Also, I think that the PR needs to be rebased on `master` and update the native image CI configuration now that #3719 has been merged

Rebase was performed locally and is building right now. But what is expect from "update the native image CI configuration"?

**Comment by geoand** [comment]
Because you have added a new integration test module, you need to add something like [this](https://github.com/quarkusio/quarkus/blob/master/azure-pipelines.yml#L286)

**Comment by misl** [comment]
Thanks, added it locally. Will soon commit and push (as local build completed successfully)

**Comment by misl** [comment]
Sorry, but I need some help on this one. I have no clue what the vertx integration test compilation errors has to do with my changes.

**Comment by geoand** [comment]
I saw the same thing earlier in a different PR - but I think it has been fixed on master since so it's probably worth rebasing again

**Comment by geoand** [comment]
Sorry for this, but it needs another rebase - there have been a lot of CI changes the last 2 days

**Comment by misl** [comment]
Help, I need help again. Is this another error due to CI changes. 

Sorry to say so, but I am staring to feel like I am wasting time on this.

**Comment by geoand** [comment]
I'm not on my machine now so I'll take a look in a few hours.

I understand the feeling, but I can assure you that you have experienced particularl bad luck having your PR arrive in the middle of large CI changes

**Comment by misl** [comment]
Let's hope this is not an bad omen ;-)

BTW regarding CI. In my opinion things are scattered all over (overstatement) the repository. Especially for extensions. I needed to make changes in various locations:
- the extension folder itself: extensions/jackson
- Integration test: integration-tests/jackson
- CI pipeline: azure-pipelines.yml
Personally I would prefer these things in a common extension root module (like extensions/jackson). The integration test could live there as well. The pipeline part can probably we auto generated (convention over configuration).

Like this there is less chance of merge conflicts. Newbies like me need less knowledge about the project structure. If (ever) extensions get separated from core Quarkus it would be a lot less work since all extension related stuff is in a single place. And possible improves development speed (less hopping around).

Sorry, this is just me always trying to improve on (working) things and bringing it rather blunt.

**Comment by geoand** [comment]
I've restarted the failing test multiple times - same result each time.
None of the other PRs are failing this test so maybe there is something wrong with this PR?

**Comment by misl** [comment]
Looking at the log I see no indication that the errors are caused by the Jackson extension. Looks more like the DynamoDB:

```
2019-08-29T11:15:10.9962932Z 11:15:09,561 INFO  [org.jbo.threads] JBoss Threads version 3.0.0.Beta5
2019-08-29T11:15:11.0949524Z [ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 3.572 s <<< FAILURE! - in io.quarkus.it.dynamodb.DynamoDbFunctionalityTest
2019-08-29T11:15:11.0961581Z [ERROR] testDynamoDbAsync  Time elapsed: 0.022 s  <<< ERROR!
2019-08-29T11:15:11.0962235Z org.junit.jupiter.api.extension.TestInstantiationException: 
2019-08-29T11:15:11.0981185Z TestInstanceFactory [io.quarkus.test.junit.QuarkusTestExtension] failed to instantiate test class [io.quarkus.it.dynamodb.DynamoDbFunctionalityTest]: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
2019-08-29T11:15:11.0981944Z 	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: Found 2 deployment problems: 
2019-08-29T11:15:11.0994909Z [1] Unsatisfied dependency for type software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient and qualifiers [@Default]
2019-08-29T11:15:11.0995959Z 	- java member: io.quarkus.it.dynamodb.DynamoDBResource#asyncClient
2019-08-29T11:15:11.0997080Z 	- declared on CLASS bean [types=[io.quarkus.it.dynamodb.DynamoDBResource, java.lang.Object], qualifiers=[@Default, @Any], target=io.quarkus.it.dynamodb.DynamoDBResource]
2019-08-29T11:15:11.1008975Z [2] Unsatisfied dependency for type software.amazon.awssdk.services.dynamodb.DynamoDbClient and qualifiers [@Default]
2019-08-29T11:15:11.1009988Z 	- java member: io.quarkus.it.dynamodb.DynamoDBResource#client
2019-08-29T11:15:11.1024370Z 	- declared on CLASS bean [types=[io.quarkus.it.dynamodb.DynamoDBResource, java.lang.Object], qualifiers=[@Default, @Any], target=io.quarkus.it.dynamodb.DynamoDBResource]
2019-08-29T11:15:11.1025035Z 	at
```

I will attempt another rebase

**Comment by misl** [comment]
Finally, all tests passed :-)

**Comment by geoand** [comment]
:+1: 

I'll check it out, but as I (think I) mentioned in a previous comment, the PR needs to be squashed to a single commit :)

**Comment by geoand** [comment]
Much better, thanks!

**Comment by FroMage** [comment]
@geoand, do you intend to review this?

**Comment by geoand** [comment]
@FroMage yup, absolutely

File: extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java
Comment (geoand) [inline extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java]: I don't think this adds much value. I would prefer it be "inlined" in it's single usage.
The change would make the style similar to what other extensions do

File: integration-tests/jackson/model/src/main/java/io/quarkus/reproducer/jacksonbuilder/model/InheritedModelWithBuilder.java
Comment (geoand) [inline integration-tests/jackson/model/src/main/java/io/quarkus/reproducer/jacksonbuilder/model/InheritedModelWithBuilder.java]: Can you please remove all `@author` tags? Quarkus doesn't use them

File: extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java
Comment (geoand) [inline extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java]: Have you explored the use of `ReflectiveHierarchyBuildItem` instead of doing this? I think it would be a very good fit for this use case

File: extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java
Comment (geoand) [inline extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java]: I am pretty sure this will fail badly when `@JsonDeserialize` is used to annotate anything but a class. Ideally we should handle all types, but maybe as a first step you can just check if the `AnnotationTarget.Kind` is a class

File: integration-tests/jackson/Readme.md
Comment (geoand) [inline integration-tests/jackson/Readme.md]: :100: on this README

File: integration-tests/jackson/Readme.md
Comment (geoand) [inline integration-tests/jackson/Readme.md]: `restful` -> `RESTful`

Comment (geoand) [reply integration-tests/jackson/Readme.md]: `post` -> `POST`

File: integration-tests/jackson/Readme.md
Comment (geoand) [inline integration-tests/jackson/Readme.md]: `unittest` -> `unit test`

File: integration-tests/jackson/Readme.md
Comment (geoand) [inline integration-tests/jackson/Readme.md]: `correct working` should probably be `correct behavior`

File: integration-tests/jackson/Readme.md
Comment (geoand) [inline integration-tests/jackson/Readme.md]: `post` -> `POST`

File: integration-tests/jackson/Readme.md
Comment (geoand) [inline integration-tests/jackson/Readme.md]: Perhaps add backticks to `201`?

File: integration-tests/jackson/Readme.md
Comment (geoand) [inline integration-tests/jackson/Readme.md]: This isn't really correct since the first way is a test of the native application, not an actual run. Would you like to rephrase this sentence?

File: integration-tests/jackson/model/src/test/java/io/quarkus/reproducer/jacksonbuilder/model/InheritedModelWithBuilderTest.java
Comment (geoand) [inline integration-tests/jackson/model/src/test/java/io/quarkus/reproducer/jacksonbuilder/model/InheritedModelWithBuilderTest.java]: :+1: from me for using `AssertJ` instead of Junit API :). Not everyone agrees, but I personally love it

File: integration-tests/jackson/service/src/main/docker/Dockerfile.jvm
Comment (geoand) [inline integration-tests/jackson/service/src/main/docker/Dockerfile.jvm]: I don't think we need the Docker files at all, can you remove them please?

**Review by geoand — CHANGES_REQUESTED** [review]
This is really good overall, great job!

I added a few comments, most of which are minor

File: extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java
Comment (misl) [reply extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java]: Replacing JSON_DESERIALIZE_FIELD_BUILDER with the string "builder" is fine with me. 

However, I prefer both DotNames as they are. For me this is about readability. In my opinion
```
index.getAnnotations(JSON_DESERIALIZE)
```
reads better than
```
index.getAnnotations(DotName.createSimple(JsonDeserialize.class.getName()))
```
Besides that, the BUILDER_VOID is inside a `for`-loop. I prefer not to rely on any compiler optimazation.

Comment (geoand) [reply extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java]: Yes, I was referring to the "builder" string only

**Comment by geoand** [comment]
Once done with the changes, please squash to a single commit.

Thanks!

**Comment by misl** [comment]
How to proceed from here? 

Do you prefer to review the individual commits or should I first squash (+ rebase) again?

**Comment by geoand** [comment]
Go ahead please with squash and rebase - the PR is not too large to require review per commit I would say

File: extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java
Comment (geoand) [reply extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java]: Sorry to trouble you, but since only the `class` case is covered can you add a `TODO` note here explaining that we should support `@JsonDeserialize` on `methods`, `fields` etc.?
That way someone eager to  contribute when looking through the code can pick this up potentially

File: integration-tests/jackson/Readme.md
Comment (geoand) [reply integration-tests/jackson/Readme.md]: You used the British English spelling :). Actually I don't think we have ever addressed this (obviously minor) issue in the guides.
So let's keep it the way you did it :)

**Comment by geoand** [comment]
Just one last uber-minor point I raised. Once it's addressed, I'll approve

File: extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java
Comment (geoand) [inline extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java]: I don't think this check is needed, it should also by of `Class` Kind.

**Review by geoand — CHANGES_REQUESTED** [review]

File: extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java
Comment (geoand) [reply extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java]: Then when that is removed, I think the `ClassInfo builderClass = combinedIndexBuildItem.getIndex().getClassByName(builderClassName);` which is used before this method is called can also be simplified

**Review by geoand — APPROVED** [review]
Thank you very much for this!

File: extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java
Comment (geoand) [inline extensions/jackson/deployment/src/main/java/io/quarkus/jackson/deployment/JacksonProcessor.java]: There is a typo in `annotationa`. Also the sentence could possible be expressed in a more clear manner, but it's definitely not something to hold up the PR for :)

**Comment by gwenneg** [comment]
Fixes #3652

**Comment by stuartwdouglas** [comment]
For some reason this PR has increased the build time of the kubernetes client native by around 40 minutes.

**Comment by geoand** [comment]
I'll check it out

**Comment by geoand** [comment]
Although I need to dig through it, my guess is that now the entire Kubernetes Client model is being registered for reflection, as opposed to being "weakly" registered.

**Comment by geoand** [comment]
I tweaked the Kubernetes extension in [this](https://github.com/quarkusio/quarkus/pull/3814) PR.

**Pull request merged**
